File: operator-api/src/routes/clipboard.js
import { Hono } from 'hono'
import { execCapture } from '../utils/exec.js'

export const clipboardRouter = new Hono()

async function wlGet() {
  const txt = await execCapture('bash', ['-lc', 'command -v wl-paste >/dev/null && wl-paste -t text || true'])
  if (txt.stdout?.length) return { type: 'text', text: txt.stdout.toString('utf8') }
  const png = await execCapture('bash', ['-lc', 'command -v wl-paste >/dev/null && wl-paste -t image/png | base64 -w0 || true'])
  if (png.stdout?.length) return { type: 'image', image_b64: png.stdout.toString('utf8'), image_mime: 'image/png' }
  return { type: 'text', text: '' }
}

async function wlSet({ type, text, image_b64, image_mime }) {
  if (type === 'text') await execCapture('bash', ['-lc', `printf %s ${JSON.stringify(text || '')} | wl-copy`])
  else if (type === 'image' && image_b64) await execCapture('bash', ['-lc', `base64 -d | wl-copy -t ${image_mime || 'image/png'}`], { input: Buffer.from(image_b64, 'base64') })
}

async function xGet() {
  const txt = await execCapture('bash', ['-lc', 'command -v xclip >/dev/null && xclip -selection clipboard -o || true'])
  return { type: 'text', text: txt.stdout.toString('utf8') }
}

async function xSet({ type, text }) {
  if (type === 'text') await execCapture('bash', ['-lc', `printf %s ${JSON.stringify(text || '')} | xclip -selection clipboard`])
}

clipboardRouter.get('/clipboard', async (c) => {
  try {
    const wayland = process.env.XDG_SESSION_TYPE === 'wayland'
    const res = wayland ? await wlGet() : await xGet()
    return c.json(res)
  } catch {
    return c.json({ type: 'text', text: '' })
  }
})

clipboardRouter.post('/clipboard', async (c) => {
  try {
    const body = await c.req.json()
    const wayland = process.env.XDG_SESSION_TYPE === 'wayland'
    if (wayland) await wlSet(body)
    else await xSet(body)
    return c.json({ ok: true })
  } catch (e) {
    return c.json({ message: String(e.message) }, 400)
  }
})

clipboardRouter.get('/clipboard/stream', async (c) => {
  // simple poller
  let lastText = ''
  const stream = new ReadableStream({
    async start(controller) {
      const enc = new TextEncoder()
      const loop = async () => {
        try {
          const res = await (process.env.XDG_SESSION_TYPE === 'wayland' ? wlGet() : xGet())
          const curr = res.type === 'text' ? res.text : res.image_b64?.slice(0, 16) || ''
          if (curr !== lastText) {
            lastText = curr
            controller.enqueue(enc.encode(`event: data\ndata: ${JSON.stringify({ ts: new Date().toISOString(), type: res.type, preview: res.type === 'text' ? (res.text || '').slice(0, 100) : 'image...' })}\n\n`))
          }
        } catch {}
        setTimeout(loop, 1000)
      }
      loop()
    },
    cancel() {}
  })
  return new Response(stream, { headers: { ...{ 'Cache-Control': 'no-cache' }, ...{ 'Content-Type': 'text/event-stream' }, 'X-SSE-Content-Type': 'application/json' } })
})

File: server-sigma/internal/routes/clipboard/clipboard.go
package clipboard

import (
	"encoding/base64"
	"encoding/json"
	"net/http"
	"os"
	"os/exec"
	"strings"
	"time"

	"kernel-operator-api/internal/utils/sse"
)

func Register(mux *http.ServeMux) {
	mux.HandleFunc("/clipboard", handleClipboard)
	mux.HandleFunc("/clipboard/stream", handleStream)
}

func have(bin string) bool {
	_, err := exec.LookPath(bin)
	return err == nil
}

func xGet() (map[string]any, error) {
	display := os.Getenv("DISPLAY")
	if display == "" {
		display = ":20"
	}
	cmd := exec.Command("bash", "-lc", "DISPLAY="+display+" xclip -selection clipboard -o 2>/dev/null || true")
	out, _ := cmd.CombinedOutput()
	return map[string]any{"type": "text", "text": string(out)}, nil
}

func wlGet() (map[string]any, error) {
	// text
	cmd := exec.Command("bash", "-lc", "wl-paste -t text 2>/dev/null || true")
	if have("wl-paste") {
		out, _ := cmd.CombinedOutput()
		if len(out) > 0 {
			return map[string]any{"type": "text", "text": string(out)}, nil
		}
		// image
		cmd = exec.Command("bash", "-lc", "wl-paste -t image/png | base64 -w0 2>/dev/null || true")
		out, _ = cmd.CombinedOutput()
		if len(out) > 0 {
			return map[string]any{"type": "image", "image_b64": string(out), "image_mime": "image/png"}, nil
		}
	}
	return map[string]any{"type": "text", "text": ""}, nil
}

func handleClipboard(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case http.MethodGet:
		var res map[string]any
		var err error
		if have("xclip") {
			res, err = xGet()
		} else {
			res, err = wlGet()
		}
		if err != nil {
			res = map[string]any{"type": "text", "text": ""}
		}
		writeJSON(w, http.StatusOK, res)
	case http.MethodPost:
		var body struct {
			Type     string `json:"type"`
			Text     string `json:"text"`
			ImageB64 string `json:"image_b64"`
			ImageMIM string `json:"image_mime"`
		}
		_ = json.NewDecoder(r.Body).Decode(&body)
		if strings.ToLower(body.Type) == "image" && body.ImageB64 != "" && have("wl-copy") {
			cmd := exec.Command("bash", "-lc", "base64 -d | wl-copy -t "+shellQuote(body.ImageMIM))
			dec, _ := base64.StdEncoding.DecodeString(body.ImageB64)
			stdin, _ := cmd.StdinPipe()
			_ = cmd.Start()
			_, _ = stdin.Write(dec)
			_ = stdin.Close()
			_ = cmd.Wait()
		} else {
			if have("xclip") {
				display := os.Getenv("DISPLAY")
				if display == "" {
					display = ":20"
				}
				_ = exec.Command("bash", "-lc", "printf %s "+shellQuote(body.Text)+" | DISPLAY="+display+" xclip -selection clipboard").Run()
			} else if have("wl-copy") {
				_ = exec.Command("bash", "-lc", "printf %s "+shellQuote(body.Text)+" | wl-copy").Run()
			}
		}
		writeJSON(w, http.StatusOK, map[string]any{"ok": true})
	default:
		w.WriteHeader(http.StatusMethodNotAllowed)
	}
}

func handleStream(w http.ResponseWriter, r *http.Request) {
	sse.Headers(w)
	flusher, _ := w.(http.Flusher)
	if flusher == nil {
		w.WriteHeader(http.StatusInternalServerError)
		return
	}
	last := ""
	t := time.NewTicker(1 * time.Second)
	defer t.Stop()
	for {
		select {
		case <-r.Context().Done():
			return
		case <-t.C:
			cur := ""
			if have("xclip") {
				m, _ := xGet()
				cur, _ = m["text"].(string)
			} else {
				m, _ := wlGet()
				if m["type"] == "text" {
					cur, _ = m["text"].(string)
				} else {
					if s, _ := m["image_b64"].(string); len(s) > 16 {
						cur = s[:16]
					}
				}
			}
			if cur != last {
				last = cur
				_ = sse.WriteJSON(w, map[string]any{
					"ts":     time.Now().UTC().Format(time.RFC3339),
					"type":   "text",
					"preview": preview(cur, 100),
				})
			}
		}
	}
}

func preview(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}

func writeJSON(w http.ResponseWriter, code int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(code)
	_ = json.NewEncoder(w).Encode(v)
}

func shellQuote(s string) string {
	return "'" + strings.ReplaceAll(s, "'", "'\\''") + "'"
}